notes/smart-pointers-2.md

```diff
@@ -0,0 +1,210 @@
+# Smart Pointers, Part 2
+
+Contributors: Edward Zhang
+
+---
+## `std::shared_ptr`
+
+Let's look at shared pointers!
+
+These are similar to unique pointers, but now we allow objects to be shared / "owned" by multiple shared pointers
+
+---
+## Reference counting
+
+If multiple shared pointers can own an object, how do we know when to `delete` the pointed-to object?
+
+Quite intuitively, we use the notion of **reference counting**. For each object, we keep a count of how many shared pointers are pointing to it
+> NOTE: this section provides some nice intuition for reference counts. However, keep reading to see how complications can arise because of the actual implementation
+
+The reference count for an object is adjusted like so:
+- When we make a new  shared pointer to the object, increment its reference count by 1
+- When a shared pointer to it is destroyed, decrement reference count by 1
+
+Note that shared pointers can be copied
+- I.e., its copy constructor and copy-assignment operators are defined
+- Copy-assignment can also increment or decrement reference counts (see next section)
+
+Finally, when a shared pointer sees that the object's reference count is 0, destroy the pointed-to object
+
+---
+## Creating shared pointers
+
+Example 1:
+```C++
+int main(){
+  int *p = new int(333); // pointing to an int object
+  std::shared_ptr<int> x(p);
+  // Now reference count to int object is 1
+
+  // y also shares ownership of int object,
+  // reference count = 2
+  std::shared_ptr<int> y(x);
+
+  // y now falls out of scope and is destroyed
+  // ref count for int object is 2-1 = 1
+
+  // x now falls out of scope and is destroyed, ref count = 0
+  // Since ref count = 0, delete int object
+
+  // Again, no memory leaks!
+}
+```
+
+Example 2: copy-assignment
+```C++
+int main(){
+  // p points to an int object, let's call the object "P"
+  int *p = new int(333);
+  std::shared_ptr<int> x(p); // ref count for P = 1
+
+  int *q = new int(124); // q points to another int object, "Q"
+  std::shared_ptr<int> y(q); // ref count for Q = 1
+
+  std::shared_ptr<int> z(x); // ref count for P = 2
+  x = y;
+  // x now goes off to share ownership of "Q", abandoning "P"
+
+  // What are the ref counts?
+  // Answer: ref count for P = 2-1 = 1, ref count for Q = 1+1 = 2
```

Comment:
This took a good minute for me to look at (Maybe because I'm not familiar with this completely), but a question I would ask here is why does z not point to q now too?